### PR TITLE
Rewrite fit_simulation.md for clarity.

### PR DESCRIPTION
**Pull request by nathanrboyer**
I am learning SciML and its related packages for the first time by going through the SciML tutorials. The `fit_simulation` was especially confusing, so I attempted to rewrite it for clarity.

### CODE REVIEW

**Comment by nathanrboyer** [comment]
I have some remaining questions.

1. I tried to change `x` and `y` to be countable numbers by multiplying them by 100 (and also changing `u0` and `pguess`), but nothing I tried would optimize close to the correct solution. I could only get my initial `pguess` to move around by order O1 not O10 or O100. How can I allow for larger state variable movement in Optimization.jl, so that the number of wolves is not 0.3?
2. The `callback` function was initially defined as a named anonymous function, and I found the syntax of that very confusing. Is there any reason to keep it that way? The [documentation ](https://docs.sciml.ai/Optimization/dev/API/solve/#Common-Solver-Options-(Solve-Keyword-Arguments)) shows `callback` defined both the original way and as a normal function. If it's all the same, I'd just use a regular function definition.
3. I could not find any documentation at all for `PolyOpt`. This tutorial should say something about why it was chosen, especially since the previous tutorial used `NLopt.LD_LBFGS` instead.

I think I understand the tutorial now, but due to the questions I still have above, I'm not confident about applying these approaches more generally. Hopefully you can review and provide some insight. Thanks.

**Comment by Vaibhavdixit02** [comment]
>I tried to change x and y to be countable numbers by multiplying them by 100 (and also changing u0 and pguess), but nothing I tried would optimize close to the correct solution. I could only get my initial pguess to move around by order O1 not O10 or O100. How can I allow for larger state variable movement in Optimization.jl, so that the number of wolves is not 0.3 but 30?

Can you show the code you ran?

>The [documentation ](https://docs.sciml.ai/Optimization/dev/API/solve/#Common-Solver-Options-(Solve-Keyword-Arguments))shows callback defined both the original way and as a normal function. If it's all the same, I'd just use a regular function definition.

Yes feel free to change that, there's no reason to prefer the anonymous function, it's just easier to write but makes sense that not for reading for beginners 😅

 >I could not find any documentation at all for PolyOpt. This tutorial should say something about why it was chosen, especially since the previous tutorial used NLopt.LD_LBFGS instead.
 
 I'll add that in, I just forgot to add it :)

**Comment by nathanrboyer** [comment]
> Can you show the code you ran?

None of the commented `pguess` vectors worked, even after 10x the original number of iterations.

```julia
using DifferentialEquations, Optimization, OptimizationPolyalgorithms, SciMLSensitivity
using ForwardDiff, Plots

# Define experimental data
t_data = 0:10
x_data = [100 277 677  97 189 610 140 134 435 325 103]
y_data = [100  26 202 191  32  63 346  51  31 455  91]
xy_data = vcat(x_data, y_data)

# Plot the provided data
scatter(t_data, xy_data', label=["x Data" "y Data"])

# Setup the ODE function
function lotka_volterra!(du, u, p, t)
    x, y = u
    α, β, δ, γ = p
    du[1] = α * x - β * x * y
    du[2] = -δ * y + γ * x * y
end

# Initial condition
u0 = xy_data[:, 1] # [100, 100]

# Simulation interval
tspan = (first(t_data), last(t_data)) # (0, 10)

# LV equation parameter. p = [α, β, δ, γ]
pguess = [100.0, 1.2, 250.0, 1.2]

# Set up the ODE problem with our guessed parameter values
prob = ODEProblem(lotka_volterra!, u0, tspan, pguess)

# Solve the ODE problem with our guessed parameter values
initial_sol = solve(prob, saveat = 1)

# View the guessed model solution
plt = plot(initial_sol, label = ["x Prediction" "y Prediction"])
scatter!(plt, t_data, xy_data', label = ["x Data" "y Data"])

# Define a loss metric function to be minimized
function loss(newp)
    newprob = remake(prob, p = newp)
    sol = solve(newprob, saveat = 1)
    loss = sum(abs2, sol .- xy_data)
    return loss, sol
end

# Define a callback function to monitor optimization progress
function callback(_, l, sol)
    display(l)
    plt = plot(sol, ylim = (0, 600), label = ["Current x Prediction" "Current y Prediction"])
    scatter!(plt, t_data, xy_data', label = ["x Data" "y Data"])
    display(plt)
    return false
end

# Set up the optimization problem with our loss function and initial guess
adtype = AutoForwardDiff()
optf = OptimizationFunction((x, _) -> loss(x), adtype)
# pguess = [1.0, 1.2, 2.5, 1.2]
# pguess = [100.0, 1.2, 250.0, 1.2]
# pguess = [1.5, 1.0, 3.0, 1.0]
# pguess = [150.0, 100.0, 300.0, 100.0]
# pguess = [150.0, 1/100, 300.0, 1/100]
pguess = [150.0, 1.0, 300.0, 1.0]
optprob = OptimizationProblem(optf, pguess)

# Optimize the ODE parameters for best fit to our data
pfinal = solve(optprob, PolyOpt(),
               callback = callback,
               maxiters = 2000)
α, β, γ, δ = round.(pfinal, digits=1)
```

> Yes feel free to change that

Okay, I have done so.

> I'll add that in

Thank you for documenting, but "Adam" and "BFGS" still mean nothing to me. Can you add a commit here about why `PolyOpt` would be chosen in this case as opposed to `NLopt.LD_LBFGS`? `NLopt.LD_LBFGS` was described as both "robust" and "performant" in the previous tutorial, so I don't understand why we switched algorithms.

**Comment by nathanrboyer** [comment]
The PR as written is accurate and works, but I will change it to use the above code instead if you know an easy way to get the optimization to converge.

**Comment by Vaibhavdixit02** [comment]
Yeah that doesn't converge for me either. I didn't look too deeply into it though, I am guessing the scaling difference shows up somewhere making things ill-conditioned. Are you finished with the all other changes then?

File: docs/src/getting_started/fit_simulation.md [hunk 1]
```diff
@@ -167,101 +215,184 @@ u0 = [1.0, 1.0]
...
-cost function. Here we'll also define a callback to monitor the solution process, more details about callbacks in Optimization.jl can be found [here](https://docs.sciml.ai/Optimization/stable/API/solve/). 
-However, this time, our function returns two things. The callback syntax is always `(value being optimized, arguments of loss return)`
-and thus this time the callback is given `(p, l, sol)`. See, returning the solution along with the loss as part of the
-loss function is useful because we have access to it in the callback to do things like plot the current solution
-against the data! Let's do that in the following way:
+The callback syntax is always:
+
+```
+callback(
+    optimization function parameters,
```
Comment (Vaibhavdixit02) [inline docs/src/getting_started/fit_simulation.md:302]:
```suggestion
    optimization variables,
```

File: docs/src/getting_started/fit_simulation.md [hunk 2]
```diff
@@ -167,101 +215,184 @@ u0 = [1.0, 1.0]
...
+    value being optimized,
```
Comment (Vaibhavdixit02) [inline docs/src/getting_started/fit_simulation.md:303]:
```suggestion
    the current loss value,
```

File: docs/src/getting_started/fit_simulation.md [hunk 3]
```diff
@@ -167,101 +215,184 @@ u0 = [1.0, 1.0]
...
+    other arguments returned from the loss function, ...
+)
+```
+
+In this case, we will provide the callback the arguments `(_, l, sol)`,
+since there are no additional optimization function parameters.
```
Comment (Vaibhavdixit02) [inline docs/src/getting_started/fit_simulation.md:309]: This is incorrect since they are the optimization variables as changed above

File: docs/src/getting_started/fit_simulation.md [hunk 4]
```diff
@@ -167,101 +215,184 @@ u0 = [1.0, 1.0]
...
-Thus every step of the optimization will show us the loss and a plot of how the solution
-looks vs the data at our current parameters.
+With this callback function, every step of the optimization will display both the loss value and a plot of how the solution compares to the training data.
+
+Now, just like [the first optimization tutorial](@ref first_opt),
+we set up our `OptimizationFunction` and `OptimizationProblem`,
+and then `solve` the `OptimizationProblem`.
+We will initialize the `OptimizationProblem` with the same `pguess` we used
+when setting up the ODE Model in Step 3.
+Observe how `Optimization.solve` brings the model closer to the experimental data as it iterates towards better ODE parameter values!
```
Comment (Vaibhavdixit02) [inline docs/src/getting_started/fit_simulation.md:336]:
```suggestion

Note that we are using the `PolyOpt()` solver choice here which is discussed https://docs.sciml.ai/Optimization/dev/optimization_packages/polyopt/ since parameter estimation of non-linear differential equations is generally a non-convex problem so we want to run a stochastic algorithm (Adam) to get close to the minimum and then finish off with a quasi-newton method (L-BFGS) to find the optima.
```

**Comment by ChrisRackauckas** [comment]
@Vaibhavdixit02 what are the remaining steps here?

**Comment by Vaibhavdixit02** [comment]
Just the one review comment left above

File: docs/src/getting_started/fit_simulation.md [hunk 5]
```diff
@@ -167,101 +215,185 @@ u0 = [1.0, 1.0]
...
+
+# View the guessed model solution
+plt = plot(initial_sol, label = ["x Prediction" "y Prediction"])
+scatter!(plt, t_data, xy_data', label = ["x Data" "y Data"])
 ```
 
-!!! note    
+Clearly the parameter values that we guessed are not correct to model this system.
+However, we can use Optimization.jl together with DifferentialEquations.jl
+to fit our parameters to our training data.
```
Comment (ChrisRackauckas) [inline docs/src/getting_started/fit_simulation.md:239]:
```suggestion
    Clearly the parameter values that we guessed are not correct to model this system.
    However, we can use Optimization.jl together with DifferentialEquations.jl
    to fit our parameters to our training data.
```

File: docs/src/getting_started/fit_simulation.md [hunk 6]
```diff
@@ -167,101 +215,185 @@ u0 = [1.0, 1.0]
...
+The sum of squares can be quickly written in Julia via `sum(abs2,x)`.
+Using this information, our loss function looks like:
 
 ```@example odefit
 function loss(newp)
     newprob = remake(prob, p = newp)
     sol = solve(newprob, saveat = 1)
-    loss = sum(abs2, sol .- data)
+    loss = sum(abs2, sol .- xy_data)
     return loss, sol
```
Comment (ChrisRackauckas) [inline docs/src/getting_started/fit_simulation.md:280]:
```suggestion
    l = sum(abs2, sol .- xy_data)
    return l, sol
```

File: docs/src/getting_started/fit_simulation.md [hunk 7]
```diff
@@ -167,101 +215,185 @@ u0 = [1.0, 1.0]
...
+```
+callback(
+    optimization variables,
+    the current loss value,
+    other arguments returned from the loss function, ...
+)
+```
+
+In this case, we will provide the callback the arguments `(_, l, sol)`,
+since there are no additional optimization function parameters.
```
Comment (ChrisRackauckas) [inline docs/src/getting_started/fit_simulation.md:309]:
```suggestion
In this case, we will provide the callback the arguments `(p, l, sol)`,
since it always takes the current state of the optimization first (`p`)
then the returns from the loss function (`l, sol`).
```

File: docs/src/getting_started/fit_simulation.md [hunk 8]
```diff
@@ -167,101 +215,186 @@ u0 = [1.0, 1.0]
...
+`Optimization.solve` will halt if/when the callback function returns `true` instead.
+Typically the `return` statement would monitor the loss value
+and stop once some criteria is reached, e.g. `return loss < 0.0001`,
+but we will stop after a set number of iterations instead.
+More details about callbacks in Optimization.jl can be found
+[here](https://docs.sciml.ai/Optimization/stable/API/solve/).
 
 ```@example odefit
-callback = function (p, l, sol)
+function callback(_, l, sol)
```
Comment (ChrisRackauckas) [inline docs/src/getting_started/fit_simulation.md:320]:
```suggestion
function callback(p, l, sol)
```

File: docs/src/getting_started/fit_simulation.md [hunk 9]
```diff
@@ -167,101 +215,186 @@ u0 = [1.0, 1.0]
...
+
+    - `u` in `OptimizationProblem` contains the state variables of `f` to be optimized.
+
+        Here: the ODE parameters `\alpha`, `\beta`, `\gamma`, and `\delta` stored in `p`
+
+    - `p` in `OptimizationProblem` contains any fixed
+    [hyperparameters](https://en.wikipedia.org/wiki/Hyperparameter_(machine_learning)) of `f`.
 
-and the answer from the optimization is our desired parameters.
+        Here: our `loss` function does not require any hyperparameters, so we pass `_` for this `p`.
```
Comment (nathanrboyer) [inline docs/src/getting_started/fit_simulation.md:400]: This line needs changed now to be consistent; `_` was changed to `p` above.

I used `_` and wrote this section in an attempt to avoid the confusion I had between this `p` and  `p = [\alpha, \beta, \gamma, \delta]` which is actually passed as `u`.

Comment (ChrisRackauckas) [reply docs/src/getting_started/fit_simulation.md:400]: And this just underwent a change anyways. @Vaibhavdixit02 can you update this as part of that whole update?

File: docs/src/getting_started/fit_simulation.md (hunk 3 above)
Comment (nathanrboyer) [reply docs/src/getting_started/fit_simulation.md]: The "variable" wording seems too vague. There are "state variables", "parameters", and "hyperparameters" which are all variables in some sense. There are two different equation contexts where those definitions change, as I tried to explain at the bottom, and that creates a lot of the confusion for this passage.

I don't really understand what is incorrect about this. `p = [\alpha, \beta, \gamma, \delta]` are not the optimization parameters in this context. There are no optimization parameters which is why passing an underscore works.

Comment (ChrisRackauckas) [reply docs/src/getting_started/fit_simulation.md]: We should probably make a table.

File: docs/src/getting_started/fit_simulation.md (hunk 9 above)
Comment (nathanrboyer) [reply docs/src/getting_started/fit_simulation.md:400]: Should instead say something like " Here: our `loss` function does not require any hyperparameters, so we do not provide the `p` argument to `OptimizationProblem`."

**Comment by nathanrboyer** [comment]
I'll also reiterate that if someone can provide rabbit/wolf population data with countable numbers that still converges under optimization, that would also make this section easier to understand. `y=0.2` wolves doesn't make any sense.
